File: code_learning/Hello_suanfa/dynamic_programming/begin_try/climbing_stairs_dp.cpp
#include <bits/stdc++.h>
using namespace std;

/* 爬楼梯：动态规划 */
int climbingStairsDP(int n)
{
    if (n == 1 || n == 2)
        return n;
    // 初始化 dp 表，用于存储子问题的解
    vector<int> dp(n + 1);
    // 初始状态：预设最小子问题的解
    dp[1] = 1;
    dp[2] = 2;
    // 状态转移：从较小子问题逐步求解较大子问题
    for (int i = 3; i <= n; i++)
        dp[i] = dp[i - 1] + dp[i - 2];

    return dp[n];
}

/* 爬楼梯：空间优化后的动态规划 */
int climbingStairsDPComp(int n)
{
    if (n == 1 || n == 2)
        return n;
    int a = 1, b = 2;
    for (int i = 3; i <= n; i++)
    {
        int tmp = b;
        b = a + b;
        a = tmp;
    }
    return b;
}

int main()
{
    int ans = climbingStairsDP(6);
    cout << ans;

    return 0;
}

File: code_learning/Hello_suanfa/dynamic_programming/begin_try/climbing_stairs_dp.py
def climbing_stairs_dp(n: int) -> int:
    """爬楼梯：动态规划"""
    if n == 1 or n == 2:
        return n;
    # 初始化 dp 表，用于存储子问题的解
    dp = [0] * (n + 1)
    # 初始状态：预设最小子问题的解
    dp[1], dp[2] = 1, 2
    # 状态转移：从较小子问题逐步求解较大子问题
    for i in range(3, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]

def climbing_stairs_dp_comp(n: int) -> int:
    """爬楼梯：空间优化后的动态规划"""
    if n == 1 or n == 2:
        return n;
    a, b = 1, 2
    for _ in range(3, n + 1):
        a, b = b, a + b    
    return b

print(climbing_stairs_dp(9))
